spec/light-client/Lightclient.tla

```diff
@@ -1,111 +1,273 @@
-  \E lastCommit \in Commits,
-     NextV \in SUBSET AllNodes \ {}:
-     \E NextVP \in [NextV -> Powers]:
-    LET new == [ height |-> height + 1, lastCommit |-> lastCommit,
-                 VP |-> last.NextVP, NextVP |-> NextVP ] IN
-    /\ IsCorrectPower(NextVP) \* the correct validators have >2/3 of power
-    /\ blockchain' = Append(blockchain, new)
-    /\ height' = height + 1                   
-
-Init ==
-  /\ height = 0
-  /\ minTrustedHeight \in 1..ULTIMATE_HEIGHT
-  (* pick a genesis block of all nodes where next correct validators have >2/3 of power *)
-  /\ \E NextV \in SUBSET AllNodes \ {}:
-       \E NextVP \in [NextV -> Powers]:
-      /\ IsCorrectPower(NextVP)
-      /\  LET VP == [n \in Corr |-> 1] 
-              genesis == [ height |-> 0, lastCommit |-> {},
-                           VP |-> VP, NextVP |-> NextVP]
+(* the events that can be issued by the lite client *)    
+OutEvents ==
+        (* request the header for a given height from the peer full node *)
+    [type: "requestHeader", height: BC!Heights]
+        \union
+        (* finish the check with a verdict *)
+    [type: "finish", verdict: BOOLEAN]  
+        \union
+    {NoEvent}      
+
+(* Produce a request event for the top element of the requestStack *)
+RequestHeaderForTopRequest(stack) ==
+    IF stack = <<>>
+    THEN NoEvent
+    ELSE  LET top == Head(stack)
+              heightToRequest ==
+                IF top.isLeft
+                THEN top.endHeight      \* the pivot is on the right
+                ELSE top.startHeight    \* the pivot is on the left
           IN
-          blockchain = <<genesis>>
+          [type |-> "requestHeader", height |-> heightToRequest]
+
+(* When starting the light client *)
+OnStart ==
+    /\ state = "init"
+    /\ inEvent.type = "start"
+    /\ state' = "working"
+        (* the block at trusted height is obtained by the user *)
+        \* TODO: the English spec does not make this explicit
+    /\ storedHeaders' = { << blockchain[TRUSTED_HEIGHT],
+                             DOMAIN blockchain[TRUSTED_HEIGHT].VP >> }
+        (* The only request on the stack ("left", h1, h2) *)
+    /\ LET initStack == << [isLeft |-> TRUE,
+                           startHeight |-> TRUSTED_HEIGHT,
+                           endHeight |-> inEvent.heightToVerify] >>
+       IN           
+        /\ requestStack' = initStack
+        /\ outEvent' = RequestHeaderForTopRequest(initStack)
+
+(* Check whether we can trust the signedHdr based on trustedHdr
+   following the trusting period method.
+   This operator is similar to CheckSupport in the English spec.
+   
+   The parameters have the following meanings:
+   - heightToTrust is the height of the trusted header
+   - heightToVerify is the height of the header to be verified
+   - trustedHdr is the trusted header (not a signed header)
+   - signedHdr is the signed header to verify (including commits)
+   *)
+CheckSupport(heightToTrust, heightToVerify, trustedHdr, signedHdr) ==
```

Comment:
I am wondering if `CheckSupport` should return a value (instead of `boolean`) that would correspond to lease duration, i.e., for how long I can still trust this header?

Reply:
The lite client could return three kinds of values: red (cannot trust), yellow (can trust, but the headers expire soon), green (can trust).